nexus/db-queries/src/db/datastore/image.rs

```diff
@@ -17,50 +19,152 @@ use crate::db::error::public_error_from_diesel_pool;
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(e, ErrorHandler::Server)
+            }),
+            false => match pagparams {
+                PaginatedBy::Id(pagparams) => paginated(
+                    project_dsl::project_image,
+                    project_dsl::id,
+                    &pagparams,
+                ),
+                PaginatedBy::Name(pagparams) => paginated(
+                    project_dsl::project_image,
+                    project_dsl::name,
+                    &pagparams.map_name(|n| Name::ref_cast(n)),
+                ),
+            }
+            .filter(project_dsl::time_deleted.is_null())
+            .filter(project_dsl::project_id.eq(authz_project.id()))
+            .select(ProjectImage::as_select())
+            .load_async::<ProjectImage>(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
+            .map(|v| v.into_iter().map(|v| v.into()).collect()),
+        }
+    }
+
+    pub async fn silo_image_list(
+        &self,
+        opctx: &OpContext,
+        authz_silo: &authz::Silo,
+        pagparams: &PaginatedBy<'_>,
+    ) -> ListResultVec<Image> {
+        opctx.authorize(authz::Action::ListChildren, authz_silo).await?;
+
+        use db::schema::silo_image::dsl;
         match pagparams {
             PaginatedBy::Id(pagparams) => {
-                paginated(dsl::image, dsl::id, &pagparams)
+                paginated(dsl::silo_image, dsl::id, &pagparams)
             }
             PaginatedBy::Name(pagparams) => paginated(
-                dsl::image,
+                dsl::silo_image,
                 dsl::name,
                 &pagparams.map_name(|n| Name::ref_cast(n)),
             ),
         }
         .filter(dsl::time_deleted.is_null())
-        .filter(dsl::project_id.eq(authz_project.id()))
-        .select(Image::as_select())
-        .load_async::<Image>(self.pool_authorized(opctx).await?)
+        .filter(dsl::silo_id.eq(authz_silo.id()))
+        .select(SiloImage::as_select())
+        .load_async::<SiloImage>(self.pool_authorized(opctx).await?)
         .await
         .map_err(|e| public_error_from_diesel_pool(e, ErrorHandler::Server))
+        .map(|v| v.into_iter().map(|v| v.into()).collect())
     }
 
-    pub async fn image_create(
+    pub async fn silo_image_create(
+        &self,
+        opctx: &OpContext,
+        authz_silo: &authz::Silo,
+        silo_image: SiloImage,
+    ) -> CreateResult<Image> {
+        let image: Image = silo_image.into();
+        opctx.authorize(authz::Action::CreateChild, authz_silo).await?;
+
+        let name = image.name().clone();
+        let silo_id = image.silo_id;
+
+        use db::schema::image::dsl;
+        let image: Image = Silo::insert_resource(
+            silo_id,
+            diesel::insert_into(dsl::image)
+                .values(image)
+                .on_conflict(dsl::id)
```

Comment:
This looks more like an "upsert" than "create".  I'm not sure if that's necessary here?

Reply:
The project images look like they're in the same boat